app/components/SessionTimeoutHandler.tsx

```diff
@@ -0,0 +1,92 @@
+import {CiipPageComponent} from 'next-env';
+import {useRouter} from 'next/router';
+import React, {useEffect, useState} from 'react';
+import LogoutWarningModal from './LogoutWarningModal';
+
+interface Props {
+  pageComponent: CiipPageComponent;
+  modalDisplaySecondsBeforeLogout: number;
+}
+
+const SessionTimeoutHandler: React.FunctionComponent<Props> = ({
+  pageComponent,
+  modalDisplaySecondsBeforeLogout
+}) => {
+  const router = useRouter();
+
+  const [showModal, setShowModal] = useState(false);
+
+  // UNIX epoch (ms)
+  const [sessionExpiresOn, setSessionExpiresOn] = useState(Infinity);
+
+  const logoutOnSessionIdled = async () => {
+    router.push({
+      pathname: '/login-redirect',
+      query: {
+        redirectTo: router.asPath,
+        sessionIdled: true
+      }
+    });
+  };
+
+  const extendSession = async () => {
+    const response = await fetch('/extend-session');
+    if (response.ok) {
+      const timeout = Number(await response.json());
+      if (timeout > modalDisplaySecondsBeforeLogout) {
+        setShowModal(false);
+      }
+      setSessionExpiresOn(timeout * 1000 + Date.now());
+    }
+  };
+
+  useEffect(() => {
+    let timeoutId;
+
+    const checkSessionIdle = async () => {
+      if (!pageComponent.isAccessProtected) return;
+
+      const response = await fetch('/session-idle-remaining-time');
+      if (response.ok) {
+        const timeout = Number(await response.json());
+
+        setSessionExpiresOn(Date.now() + timeout * 1000);
+
+        if (timeout > modalDisplaySecondsBeforeLogout) {
+          setShowModal(false);
+          timeoutId = setTimeout(() => {
+            checkSessionIdle();
+          }, (timeout - modalDisplaySecondsBeforeLogout) * 1000);
+        } else if (timeout > 0) {
+          setShowModal(true);
+          timeoutId = setTimeout(() => {
+            checkSessionIdle();
+          }, timeout * 1000);
+        } else {
+          logoutOnSessionIdled();
+        }
+      }
+    };
+
+    checkSessionIdle();
+
+    // Return cleanup function
+    return () => {
+      clearTimeout(timeoutId);
+    };
+  }, [router, pageComponent]);
+
+  return (
+    showModal && (
+      <LogoutWarningModal
+        inactivityDelaySeconds={modalDisplaySecondsBeforeLogout}
+        expiresOn={sessionExpiresOn}
+        onExtendSession={() => {
+          extendSession();
+        }}
```

Comment:
```suggestion
        onExtendSession={extendSession}
```
unless there's a reason for the anonymous arrow function here?

Reply:
Just my react reflex to avoid the `bind()` call but it doesn't matter here :)